src/ProxyClient/AbstractProxyClient.php

```diff
@@ -120,196 +70,84 @@ protected function getBaseUrl()
-     * Requests are sent in parallel to minimise impact on performance.
+     * Initialize the request queue
      *
-     * @param RequestInterface[] $requests Requests
-     *
-     * @throws ExceptionCollection
+     * @param array  $servers
+     * @param string $baseUri
      */
-    private function sendRequests(array $requests)
+    protected function initQueue(array $servers, $baseUri)
     {
-        $allRequests = array();
-
-        foreach ($requests as $request) {
-            $headers = $request->getHeaders()->toArray();
-            // Force to re-create Host header if empty, as Apache chokes on this. See #128 for discussion.
-            if (empty($headers['Host'])) {
-                unset($headers['Host']);
-            }
-            foreach ($this->servers as $server) {
-                $proxyRequest = $this->createRequest(
-                    $request->getMethod(),
-                    $server.$request->getResource(),
-                    $headers
-                );
-                $allRequests[] = $proxyRequest;
-            }
-        }
-
-        try {
-            $this->client->send($allRequests);
-        } catch (GuzzleExceptionCollection $e) {
-            $this->handleException($e);
-        }
+        $this->queue = new RequestQueue($servers, $baseUri);
     }
 
     /**
-     * Handle request exception
-     *
-     * @param GuzzleExceptionCollection $exceptions
+     * @param ResponseInterface[] $responses
      *
-     * @throws ExceptionCollection
+     * @return ProxyResponseException[]
      */
-    protected function handleException(GuzzleExceptionCollection $exceptions)
+    private function handleErrorResponses(array $responses)
     {
-        $collection = new ExceptionCollection();
+        $exceptions = [];
 
-        foreach ($exceptions as $exception) {
-            if ($exception instanceof CurlException) {
-                // Caching proxy unreachable
-                $e = ProxyUnreachableException::proxyUnreachable(
-                    $exception->getRequest()->getHost(),
-                    $exception->getMessage(),
-                    $exception->getRequest()->getRawHeaders(),
-                    $exception
-                );
-            } elseif ($exception instanceof RequestException) {
-                // Other error
-                $e = ProxyResponseException::proxyResponse(
-                    $exception->getRequest()->getHost(),
-                    $exception->getCode(),
-                    $exception->getMessage(),
-                    $exception->getRequest()->getRawHeaders(),
-                    $exception
-                );
-            } else {
-                // Unexpected exception type
-                $e = $exception;
+        foreach ($responses as $response) {
+            if ($response->getStatusCode() >= 400
+                && $response->getStatusCode() < 600
+            ) {
+                $exceptions[] = ProxyResponseException::proxyResponse($response);
```

Comment:
Now throwing exceptions again for responses with error status codes. Unfortunately, I see no easy way to get the corresponding request. Although Guzzle should return responses in the same order as the requests, things get difficult when the responses are mixed with exceptions (for timeout errors).